templates/activity-stream/impression-stats/impression-stats.1.schema.json

```diff
@@ -0,0 +1,87 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "type": "object",
+  "title": "impression-stats",
+  "properties": {
+    "action": {
+      "type": "string",
+      "pattern": "^activity_stream_impression_stats$"
+    },
+    "client_id": {
+      "type": "string",
+      "pattern": "^n/a$"
+    },
+    "session_id": {
+      "type": "string",
+      "pattern": "^n/a$"
+    },
```

Comment:
These first three fields are all expected to have static values, so we check those via regex patterns, but don't include the fields in the "required" section below. That means we will validate payloads either missing the "action" field or having action set to "activity_stream_impression_stats" exactly.

The idea is that this provides us compatibility with existing payloads today, but would allow us to stop sending these static fields in some future release where we stop needing to support sending payloads via the ping-centre infrastructure.

Reply:
`action` is merely a ping-centre specific field as the ping identifier, and we don't even bother storing it in the database, so I think we can leave it out here.

Both `client_id` and `session_id` are also omitable, they are all set as `n/a` in the current A-S payload for backward compatibility. We don't need to bear this burden in the new schema.

As a follow up question, how would the new data pipeline process those fields that are not defined in the schema? Does it simply throw them away, or still pack and store them in the database?

Reply:
> how would the new data pipeline process those fields that are not defined in the schema? Does it simply throw them away, or still pack and store them in the database?

In the existing AWS pipeline, extra fields are thrown away. We have a design goal for GCP of packaging up any remaining fields not included in the schema as a JSON blob, and including that as an additional field in the destination BigQuery table. Does that behavior sound good to you?

The code for this has not been written yet, though. I'll do a little more digging into what would be necessary to provide that.

Reply:
Yep, that'll be great!

Also, we will make sure that anything unnecessary won't be included in the payload.